include/aspect/melt.h

```diff
@@ -0,0 +1,329 @@
+                                                    internal::Assembly::Scratch::StokesSystem<dim>  &scratch,
+                                                    internal::Assembly::CopyData::StokesSystem<dim> &data);
+    }
+  }
+
+
+  /**
+   * Class to contain all runtime parameters and other helper functions
+   * related to melt transport. A global instance can be retrieved with
+   * MeltHandler<dim>::get(), but keep in mind that it only exists if
+   * parameters.include_melt_transport is true.
+   */
+  template <int dim>
+  class MeltHandler: public SimulatorAccess<dim>
+  {
+    public:
+      MeltHandler(ParameterHandler &prm);
+
+      /**
+       * Declare additional parameters that are needed in models with
+       * melt transport (including the fluid pressure boundary conditions).
+       */
+      static void declare_parameters (ParameterHandler &prm);
+
+      /**
+       * Parse additional parameters that are needed in models with
+       * melt transport (including the fluid pressure boundary conditions).
+       *
+       * This has to be called before edit_finite_element_variables,
+       * so that the finite elements that are used for the additional melt
+       * variables can be specified in the input file and are parsed before
+       * the introspection object is created.
+       */
+      void parse_parameters (ParameterHandler &prm);
+
+      /**
+       * Add the additional variables we need in simulations with melt
+       * migration to the list of variables, which will be used later
+       * to set up the introspection object.
+       */
+      void edit_finite_element_variables(const Parameters<dim> &parameters, std::vector<VariableDeclaration<dim> > &variables);
+
+      /**
+       * Setup SimulatorAccess for the plugins related to melt transport.
+       */
+      void initialize_simulator (const Simulator<dim> &simulator_object);
+
+      /**
+       * Compute fluid velocity and solid pressure in this ghosted solution vector.
+       * @param solution
+       */
+      void compute_melt_variables(LinearAlgebra::BlockVector &solution);
+
+      /**
+       * Return whether this object refers to the porosity field.
+       */
+      bool is_porosity (const typename Simulator<dim>::AdvectionField &advection_field) const;
+
+      /**
+       * The porosity limit for melt migration. For smaller porosities, the equations
+       * reduce to the Stokes equations and neglect melt transport. In practice, this
+       * means that all terms in the assembly related to the migration of melt are set
+       * to zero for porosities smaller than this threshold.
+       * This does not include the compaction term $p_c/\xi$, which is necessary for the
+       * solvability of the linear system, but does not influence the solution variables
+       * of the Stokes problem (in the absence of porosity).
+       */
+      double melt_transport_threshold;
```

Comment:
I don't think so... we typically need them in the material model/heating model/assembly.